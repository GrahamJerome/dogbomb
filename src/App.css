/*! APP.CSS */

/* APPLICATION STYLES
// ========================================================
//
// Table of Contents:
//
//  1. App / Main
//  00. Misc
//  000. Responsive / Overrides
*/

/* 1. App / Main
/ -------------------------------------------------------- */

body {
  background-size: cover;
  height: 100vh;
  width: 100vw;
}

.cody-bomb {
  background: url(/images/cody-1.png) right bottom no-repeat transparent;
  background-size: contain;
  bottom: 0;
  height: 508px;
  position: absolute;
  right: 0;
  width: 490px;
  z-index: 999;
}

/* 00. Misc
/ -------------------------------------------------------- */

/* 000. Responsive Overrides
/ -------------------------------------------------------- */