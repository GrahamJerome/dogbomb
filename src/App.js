import React, {Component} from 'react';
import * as UnsplashAPI from './UnsplashAPI'
import './App.css';

class App extends Component {
  // Sets a default values on first load.
  state = {
    dogImageUrl: 'https://i.imgur.com/iOvgJti.png',
    imageQuery: 'wedding'
  }

  // Fire up the first random image on component mount.
  componentDidMount() {
    document.addEventListener("DOMContentLoaded", () => {
      this.updateBgImage()
      this.updateDogImage()
    })
  }

  // Use the UnsplashAPI to query the random image, return the image url,
  // and adjusts the body background image style tag.
  updateBgImage = () => {
    console.log('Query Unsplash API for: ' + this.state.imageQuery)
    UnsplashAPI.getRandom(this.state.imageQuery)
      .then(imageUrl => {
        document.body.style.backgroundImage = `url(${imageUrl})`
      })
      .catch(err => console.log(err))
  }

  updateDogImage = () => {
    console.log('@@@' + document.getElementById('dog-bomb'))
    console.log('Loading dog imageUrl: ' + this.state.dogImageUrl)
    // the url for the dog image should be cached from the http.send() request.
    document.body.querySelector('.dog-bomb').style.backgroundImage = `url(${this.state.dogImageUrl})`
  }

  // When a search is triggered, it triggers an update of the background and dog image.
  // I have allowed the same query to be searched multiple times as it
  // will probably produce different results each query.
  handleSearchSubmit = (event) => {
    event.preventDefault();

    let imageQuery = document.body.querySelector("input[name='image-query']").value
    this.setState({imageQuery: imageQuery})
    this.updateBgImage()

    let dogImageUrl = document.body.querySelector("input[name='dog-image-url']").value
    if (dogImageUrl !== this.state.dogImageUrl && this.validImageUrl(dogImageUrl)) {
      this.setState({dogImageUrl: dogImageUrl})
      this.updateDogImage()
    }
  }

  validImageUrl = (imageUrl) => {
    const validTypes = ['.png','.jpg','.jpeg','.gif','.tiff']
    let ext = imageUrl.substr(imageUrl.lastIndexOf('.'))

    if (!validTypes.includes(ext)) {
      alert('dog image not valid image file type')
      return false
    }

    let http = new XMLHttpRequest();

    http.open('HEAD', imageUrl, false)

    try {
      http.send()
    }
    catch(err) {
      console.log(err)
      alert('dog image url not valid')
      return false
    }
  }

  // Update the imageQuery state value as the user types.
  handleQueryChange = (event) => {
    this.setState({imageSearchValue: event.target.value})
  }

  render() {
    return (
      <div>
        <form className="dog-bomb-search form-inline p-3 bg-white" onSubmit={this.handleSearchSubmit}>
          <div className="form-group mb-2 mb-md-0">
            <label htmlFor="dog-image-url">Dog Image Url</label>
            <input type="url" className="form-control mx-3" name="dog-image-url" defaultValue={this.state.dogImageUrl} placeholder="Dog image url" />
          </div>
          <div className="form-group mb-2 mb-md-0">
            <label htmlFor="image-query">Category</label>
            <input type="text" className="form-control mx-3" name="image-query" defaultValue={this.state.imageQuery} placeholder="Image category" />
          </div>
          <button type="submit" className="btn btn-primary mb-2">Bombs away!</button>
        </form>
        <div className='dog-bomb'></div>
      </div>
    );
  }
}

export default App;
