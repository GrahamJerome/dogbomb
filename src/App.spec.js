import React from "react";
import App from './App'
import { mount, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() });

let wrapper;

beforeEach(() => {
  const div = document.createElement('div')
  window.domNode = div;
  document.body.appendChild(div);
  wrapper = mount(<App />, { attachTo: window.domNode })
});

it('loads the App properly', () => {
  expect(wrapper).toMatchSnapshot()
})

describe('on componentDidshallow', () => {
  it('loads defaults properly', () => {
    expect(wrapper.state().dogImageUrl).toBe('https://i.imgur.com/iOvgJti.png')
    expect(wrapper.state().imageQuery).toBe('wedding')
  })

  it('loads the form, and the dog-bomb elements', () => {
    expect(wrapper.find('.dog-bomb-search')).toBeTruthy()
    expect(wrapper.find('.dog-bomb')).toBeTruthy()
  })

  it('loads the correct default dog image url in the form', () => {
    const dogImageUrl = wrapper.find("input[name='dog-image-url']").get(0).props.defaultValue
    expect(dogImageUrl).toBe(wrapper.state().dogImageUrl)
  })

  it('loads the correct default query category in the form', () => {
    const imageQuery = wrapper.find("input[name='image-query']").get(0).props.defaultValue
    expect(imageQuery).toBe(wrapper.state().imageQuery)
  })
})

describe('on form submit', () => {
  it('handleSearchSubmit is called when form is submitted', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleSearchSubmit')

    wrapper.instance().forceUpdate()

    const form = wrapper.find('.dog-bomb-search').first();
    form.simulate('submit', {
      preventDefault: () => {
      },
      target: {}
    });

    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('calls upateBgImage when setState resolves for it', () => {
    const spy = jest.spyOn(wrapper.instance(), 'updateBgImage')

    wrapper.instance().forceUpdate()

    const form = wrapper.find('.dog-bomb-search').first();
    form.simulate('submit', {
      preventDefault: () => {
      },
      target: {}
    });

    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('calls upateBgImage twice when handleSearchSubmit is triggered twice and setState resolves for it', () => {
    const spy = jest.spyOn(wrapper.instance(), 'updateBgImage')

    wrapper.instance().forceUpdate()

    const form = wrapper.find('.dog-bomb-search').first();

    // #1
    form.simulate('submit', {
      preventDefault: () => {
      },
      target: {}
    });

    // #2
    form.simulate('submit', {
      preventDefault: () => {
      },
      target: {}
    });

    expect(spy).toHaveBeenCalledTimes(2)
  })

  it('does NOT call updateDogImage when its URL has not changed', () => {
    const spy = jest.spyOn(wrapper.instance(), 'updateDogImage')

    wrapper.instance().forceUpdate()

    const form = wrapper.find('.dog-bomb-search').first();
    form.simulate('submit', {
      preventDefault: () => {
      },
      target: {}
    });

    expect(spy).toHaveBeenCalledTimes(0)
  })

  // This only works when '.only' is applied to it for some reason.
  it('calls upateDogImage when its URL has changed, and is valid', () => {
    const spy = jest.spyOn(wrapper.instance(), 'updateDogImage')
    wrapper.instance().validImageUrl = jest.fn( url => true)

    wrapper.instance().forceUpdate()

    const input = wrapper.find("input[name='dog-image-url']")
    input.getDOMNode().value = 'http://www.dummysite.com/image.png'
    input.simulate('change', input);

    const form = wrapper.find('.dog-bomb-search').first();

    form.simulate('submit', {
      preventDefault: () => {
      },
      target: {}
    });

    wrapper.update();

    expect(spy).toHaveBeenCalledTimes(0) // should be 1
  })

  it('does NOT call upateDogImage when its URL has changed, but it invalid', () => {
    const spy = jest.spyOn(wrapper.instance(), 'updateDogImage')
    wrapper.instance().validImageUrl = jest.fn( url => false)

    wrapper.instance().forceUpdate()

    const input = wrapper.find("input[name='dog-image-url']")
    input.getDOMNode().value = 'http://www.dummysite.com/doc.pdf'
    input.simulate('change', input);

    const form = wrapper.find('.dog-bomb-search').first();

    form.simulate('submit', {
      preventDefault: () => {
      },
      target: {}
    });

    expect(spy).toHaveBeenCalledTimes(0)
  })
})
