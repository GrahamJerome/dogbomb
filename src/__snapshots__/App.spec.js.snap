// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads the App properly 1`] = `ReactWrapper {}`;
